File: crates/chameleon/cpp/locale.cpp
#include "chameleon/cpp/icu_includes.h"
#include "chameleon/cpp/locale.h"
#include "chameleon/cpp/enums.h"
#include "chameleon/src/locale.rs.h"

rust::Vec<LocaleInformation> get_locale_information(OutputFor output_for) {
    rust::Vec<LocaleInformation> locale_info;

    int32_t count;
    const icu::Locale* list = NULL;
    icu::UnicodeString result;

    switch (output_for) {
        case OutputFor::AllLocales: list = icu::Locale::getAvailableLocales(count); break;
        case OutputFor::NumberFormatLocales: list = icu::NumberFormat::getAvailableLocales(count); break;
        case OutputFor::DateFormatLocales: list = icu::DateFormat::getAvailableLocales(count); break;
        default: return locale_info;
    }

    std::string tmp;
    LocaleInformation locale_info_item;
    for (int i = 0; i < count; i++) {
        locale_info_item.locale = &list[i];
        locale_info_item.locale_name = list[i].getName();

        result = ""; tmp = "";
        list[i].getDisplayName(icu::Locale::getUS(), result);
        locale_info_item.locale_display_name = result.toUTF8String(tmp);

        locale_info.push_back(locale_info_item);
    }

    return locale_info;
}

File: crates/chameleon/src/locale.rs
#[cxx::bridge]
mod ffi_icu {
    unsafe extern "C++" {
        include!("chameleon/cpp/icu_includes.h");

        type Locale;

    }

    #[derive(Copy)]
    struct LocaleInformation {
        locale: *const Locale,
        locale_name: String,
        locale_display_name: String,
    }

    unsafe extern "C++" {
        include!("chameleon/cpp/locale.h");

        type OutputFor = crate::format::OutputFor;
        fn get_locale_information(output_for: OutputFor) -> Vec<LocaleInformation>;
    }
}

pub use ffi_icu::*;

